cpp/Esri/ArcGISRuntime/Toolkit/AuthenticationController.cpp

```diff
@@ -0,0 +1,345 @@
+{
+  if (!m_currentChallenge)
+    return;
+
+  auto c = m_currentChallenge;
+  cleanup();
+  c->continueWithUsernamePassword(username, password);
+}
+
+/*!
+ \brief Calls \c continueWithOAuthAuthorizationCode on the current challenge.
+ 
+ \list
+ \li \a oAuthAuthorizationCode Authorization code.
+ \endlist
+ */
+void AuthenticationController::continueWithOAuthAuthorizationCode(const QString& oAuthAuthorizationCode)
+{
+  if (!m_currentChallenge)
+    return;
+
+  auto c = m_currentChallenge;
+  cleanup();
+  c->continueWithOAuthAuthorizationCode(oAuthAuthorizationCode);
+}
+
+/*!
+ \brief Calls \c continueWithClientCertificate on the current challenge.
+ 
+ \list
+ \li \a clientCertificateIndex The index is the index of the certificate
+     as seen in \l clientCertificateInfos.
+ \endlist
+ */
+void AuthenticationController::continueWithClientCertificate(int clientCertificateIndex)
+{
+  if (!m_currentChallenge)
+    return;
+
+  auto c = m_currentChallenge;
+  cleanup();
+  c->continueWithClientCertificate(clientCertificateIndex);
+}
+
+/*!
+ \brief Calls \c continueWithSslHandshake on the current challenge.
+ 
+ \list
+ \li \a trust When true, trusts the misconfigured resource.
+ \li \a remember Remember the user's preferences this session.
+ \endlist
+ */
+void AuthenticationController::continueWithSslHandshake(bool trust, bool remember)
+{
+  if (!m_currentChallenge)
+    return;
+
+  auto c = m_currentChallenge;
+  cleanup();
+  c->continueWithSslHandshake(trust, remember);
+}
+
+/*!
+ \brief Calls \c cancel on the current challenge.
+ */
+void AuthenticationController::cancel()
+{
+  if (!m_currentChallenge)
+    return;
+
+  auto c = m_currentChallenge;
+  cleanup();
+  c->cancel();
+}
+
+/*!
+ \internal
+ \brief Clears the reference to the current challenge. Also deletes the current
+ challenge where applicable.
+ */
+void AuthenticationController::cleanup()
+{
+  // This controller has the option of
+  // "owning" the challenge and may delete
+  // it after use.
+  if (m_deleteChallengeOnProcessed) 
+    m_currentChallenge->deleteLater();
+
+  m_currentChallenge = nullptr;
+  emit challengeChanged();
+}
+
+/*!
+  \fn void Esri::ArcGISRuntime::Toolkit::AuthenticationController::challengeChanged()
+  \brief Emitted when the challenge changes.
+ */
+
+/*!
+  \fn void Esri::ArcGISRuntime::Toolkit::AuthenticationController::clientCertificateInfosChanged()
+  \brief Emitted when the clientCertificateInfos updates.
+  \sa clientCertificateInfosChanged
+ */
+
+/*!
+  \fn void Esri::ArcGISRuntime::Toolkit::AuthenticationController::clientCertificatePasswordRequired(QUrl certificate)
+  \brief Emitted when a \a certificate that was added to the AuthenticationManager requires a password.
+  \sa addClientCertificate
+ */
+
+/*!
+  \property Esri::ArcGISRuntime::Toolkit::AuthenticationController::currentChallengeUrl
+  \brief Url of current challenge.
```

Comment:
I'd recommend using some articles to make more complete sentences - 

```suggestion
  \brief The Url of the current challenge.
```

The next couple of descriptions have this styling as well